src/pygeoprocessing/routing/routing.pyx

```diff
@@ -4130,49 +4135,95 @@ def detect_outlets(d8_flow_dir_raster_path_band, target_outlet_vector_path):
         os.path.splitext(target_outlet_vector_path)[0])
     outlet_layer = outlet_vector.CreateLayer(
         outet_basename, raster_srs, ogr.wkbPoint)
-    # i and j indicate the coordinates of the point in raster space wereas
+    # i and j indicate the coordinates of the point in raster space whereas
     # the geometry is in projected space
     outlet_layer.CreateField(ogr.FieldDefn('i', ogr.OFTInteger))
     outlet_layer.CreateField(ogr.FieldDefn('j', ogr.OFTInteger))
     outlet_layer.CreateField(ogr.FieldDefn('ID', ogr.OFTInteger))
     outlet_layer.StartTransaction()
 
     cdef time_t last_log_time = ctime(NULL)
-    # iterate by iterblocks so ManagedRaster can efficiently cache reads
+    # iterate by iterblocks so ReadAsArray can efficiently cache reads
     # and writes
+    LOGGER.info('outlet detection: 0% complete')
     for block_offsets in pygeoprocessing.iterblocks(
             d8_flow_dir_raster_path_band, offset_only=True):
         xoff = block_offsets['xoff']
         yoff = block_offsets['yoff']
         win_xsize = block_offsets['win_xsize']
         win_ysize = block_offsets['win_ysize']
-        for yi in range(win_ysize):
-            yi_root = yi+yoff
+
+        # Make an array with a 1 pixel border around the iterblocks window
+        # That border will be filled in with nodata or data from the raster
+        # if the window does not align with a top/bottom/left/right edge
+        flow_dir_block = numpy.empty(
+            (win_ysize+2, win_xsize+2), dtype=numpy.uint8)
+
+        # Test for left border and if so stripe nodata on the left margin
+        x_off_border = 0
+        if xoff > 0:
+            x_off_border = 1
+        else:
+            flow_dir_block[:, 0] = flow_dir_nodata
+
+        # Test for top border and if so stripe nodata on the top margin
+        y_off_border = 0
+        if yoff > 0:
+            y_off_border = 1
+        else:
+            flow_dir_block[0, :] = flow_dir_nodata
+
+        # Test for right border and if so stripe nodata on the right margin
+        win_xsize_border = 0
+        if xoff+win_xsize < raster_x_size-1:
+            win_xsize_border += 1
+        else:
+            flow_dir_block[:, -1] = flow_dir_nodata
+
+        # Test for bottom border and if so stripe nodata on the bottom margin
+        win_ysize_border = 0
+        if yoff+win_ysize < raster_y_size-1:
+            win_ysize_border += 1
+        else:
+            flow_dir_block[-1, :] = flow_dir_nodata
+
+        # Read iterblock plus a possible margin on top/bottom/left/right side
+        d8_flow_dir_band.ReadAsArray(

```

Comment:
Do you lose the efficiency of reading by `iterblocks` blocks when you add a 1 pixel border, since the dimensions now cover multiple memory blocks? I did the same thing in delineateit and I don't know if there's any way around it.

Reply:
No it's still pretty fine. GDAL can cache many memory blocks so when you hit the border it'll have those cached and the next window you slide over it won't need to re-load them. Also if you make `iterblocks` operate over many blocks with generous `largest_block` you can minimize that cost even more. In practice I haven't seen it be anything significant enough to measure or worry about.

Reply:
neat!